### ISSUE TITLE
Dynamic Training Pipeline

### ISSUE THREAD

**Issue by iluveu28** [dotnet/machinelearning#5637]
### System information

- **OS version/distro**: Win x64 and Linux x64
- **.NET Version (eg., dotnet --info)**: 3.1 and 5.0

### Issue

I create a web api application that accepts a csv file. I want to train and predict without hardcoding any fields coz the user can upload any csv file with any header names they desire. We can assume the last column is the label.

Something like how your tutorial here generates the codes on the fly - https://dotnet.microsoft.com/learn/ml-dotnet/get-started-tutorial/create

I'm OK to fix the algo & assume the last column is the label. However, How can I pick the feature columns, one hot encoding and one hot hash encoding columns on the fly?

### Source code / logs

Example with LigthtGbm algo :-
----------------------------------

            var dataProcessPipeline = mlContext.Transforms.Conversion.MapValueToKey("BinRootCause", "BinRootCause")
                                      .Append(mlContext.Transforms.Categorical.OneHotEncoding(new[] { new InputOutputColumnPair("Lot", "Lot"), new InputOutputColumnPair("Program", "Program"), new InputOutputColumnPair("SBLTriggeredBin", "SBLTriggeredBin"), new InputOutputColumnPair("SBLTriggerRule", "SBLTriggerRule"), new InputOutputColumnPair("WW", "WW"), new InputOutputColumnPair("DevRevStep", "DevRevStep"), new InputOutputColumnPair("SummaryId", "SummaryId"), new InputOutputColumnPair("Product", "Product"), new InputOutputColumnPair("Family", "Family"), new InputOutputColumnPair("TIU_SOC", "TIU_SOC"), new InputOutputColumnPair("TESTER_SOC_TH", "TESTER_SOC_TH") }))
                                      .Append(mlContext.Transforms.Categorical.OneHotHashEncoding(new[] { new InputOutputColumnPair("Operation", "Operation"), new InputOutputColumnPair("TestedQty", "TestedQty"), new InputOutputColumnPair("TestEndDate", "TestEndDate"), new InputOutputColumnPair("InterfaceBin", "InterfaceBin"), new InputOutputColumnPair("FunctionalBin", "FunctionalBin") }))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("VisualId_tf", "VisualId"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("SortLot_tf", "SortLot"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TESTER_CELL_tf", "TESTER_CELL"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TESTER_tf", "TESTER"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TIU_tf", "TIU"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TH_tf", "TH"))
                                      .Append(mlContext.Transforms.Concatenate("Features", new[] { "Lot", "Program", "SBLTriggeredBin", "SBLTriggerRule", "WW", "DevRevStep", "SummaryId", "Product", "Family", "TIU_SOC", "TESTER_SOC_TH", "Operation", "TestedQty", "TestEndDate", "InterfaceBin", "FunctionalBin", "VisualId_tf", "SortLot_tf", "TESTER_CELL_tf", "TESTER_tf", "TIU_tf", "TH_tf", "SBLValue", "SBLLimit", "SortX", "SortY", "SortWafer" }));
            // Set the training algorithm 
            var trainer = mlContext.MulticlassClassification.Trainers.LightGbm(labelColumnName: @"BinRootCause", featureColumnName: "Features")
                                      .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel", "PredictedLabel"));

Example with FastTree algo :
------------------------------

            var dataProcessPipeline = mlContext.Transforms.Conversion.MapValueToKey("BinRootCause", "BinRootCause")
                                      .Append(mlContext.Transforms.Categorical.OneHotEncoding(new[] { new InputOutputColumnPair("Lot", "Lot"), new InputOutputColumnPair("Program", "Program"), new InputOutputColumnPair("SBLTriggeredBin", "SBLTriggeredBin"), new InputOutputColumnPair("SBLTriggerRule", "SBLTriggerRule"), new InputOutputColumnPair("WW", "WW"), new InputOutputColumnPair("DevRevStep", "DevRevStep"), new InputOutputColumnPair("SummaryId", "SummaryId"), new InputOutputColumnPair("Product", "Product"), new InputOutputColumnPair("Family", "Family"), new InputOutputColumnPair("TIU_SOC", "TIU_SOC"), new InputOutputColumnPair("TESTER_SOC_TH", "TESTER_SOC_TH") }))
                                      .Append(mlContext.Transforms.Categorical.OneHotHashEncoding(new[] { new InputOutputColumnPair("Operation", "Operation"), new InputOutputColumnPair("TestedQty", "TestedQty"), new InputOutputColumnPair("TestEndDate", "TestEndDate"), new InputOutputColumnPair("InterfaceBin", "InterfaceBin"), new InputOutputColumnPair("FunctionalBin", "FunctionalBin") }))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("VisualId_tf", "VisualId"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("SortLot_tf", "SortLot"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TESTER_CELL_tf", "TESTER_CELL"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TESTER_tf", "TESTER"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TIU_tf", "TIU"))
                                      .Append(mlContext.Transforms.Text.FeaturizeText("TH_tf", "TH"))
                                      .Append(mlContext.Transforms.Concatenate("Features", new[] { "Lot", "Program", "SBLTriggeredBin", "SBLTriggerRule", "WW", "DevRevStep", "SummaryId", "Product", "Family", "TIU_SOC", "TESTER_SOC_TH", "Operation", "TestedQty", "TestEndDate", "InterfaceBin", "FunctionalBin", "VisualId_tf", "SortLot_tf", "TESTER_CELL_tf", "TESTER_tf", "TIU_tf", "TH_tf", "SBLValue", "SBLLimit", "SortX", "SortY", "SortWafer" }))
                                      .AppendCacheCheckpoint(mlContext);
            // Set the training algorithm 
            var trainer = mlContext.MulticlassClassification.Trainers.OneVersusAll(mlContext.BinaryClassification.Trainers.FastTree(labelColumnName: @"BinRootCause", featureColumnName: "Features"), labelColumnName: @"BinRootCause")
                                      .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel", "PredictedLabel"));

**Comment by iluveu28** [comment]
@antoniovs1029 could you advice?

**Comment by michaelgsharp** [comment]
So the sample you linked for ModelBuilder actually generates code files based on the input. Because each input column can have different types and be used for different things, its actually a fairly complex thing to do. I would recommend looking into something like [AutoML](https://docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/how-to-use-the-automl-api) which should be able to do basically what you are suggesting. Otherwise, you would have to build all that logic out yourself.

**Comment by iluveu28** [comment]
So I followed your AutoML sample here - https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.automl.multiclassclassificationexperiment?view=ml-dotnet-preview

However, I do not want to hard code the columns of the csv file with a class. So I use the InferColumns method below.

            var columnInference = _mlContext.Auto().InferColumns(trainingFilePath, labelColumnName);
            var textLoader = _mlContext.Data.CreateTextLoader(columnInference.TextLoaderOptions);

            var trainingDataView = textLoader.Load(trainingFilePath);

In the training data csv input file I got 27 columns below :-

string[] inputColNames = { "Lot", "Operation", "Program", "SBLTriggeredBin", "SBLTriggerRule", "TestedQty", "SBLValue", "SBLLimit", "TestEndDate", "SortX", "SortY", "InterfaceBin", "FunctionalBin", "VisualId", "WW", "DevRevStep", "SummaryId", "Product", "Family", "SortLot", "SortWafer", "TIU_SOC", "TESTER_SOC_TH", "TESTER_CELL", "TESTER", "TIU", "TH" };

But when I inspect the trainingDataView, I only see 13 columns in the model schema and there's a new column generated called Features with a Vector type. I'm able to save this model into a zip file so no issue with training.

![image](https://user-images.githubusercontent.com/2324842/113500162-6a662980-954e-11eb-80cf-e88fe5ba10b4.png)

The problem arises when I tried to retrieve the model to predict. Here's a snippet of the predict method :-

                string[] inputColNames = { "Lot", "Operation", "Program", "SBLTriggeredBin", "SBLTriggerRule", "TestedQty", "SBLValue", "SBLLimit", "TestEndDate", "SortX", "SortY", "InterfaceBin", "FunctionalBin", "VisualId", "WW", "DevRevStep", "SummaryId", "Product", "Family", "SortLot", "SortWafer", "TIU_SOC", "TESTER_SOC_TH", "TESTER_CELL", "TESTER", "TIU", "TH" };

                string labelColumnName = "BinRootCause";

                TextLoader.Column[] columns = new TextLoader.Column[inputColNames.Length + 1];
                int colCount = 0;
                foreach (string colName in inputColNames)
                {
					if (colName == "Operation" || colName == "TestedQty" || colName == "SBLValue" || colName == "SBLLimit" || colName == "SortX" || colName == "SortY" || colName == "InterfaceBin" || colName == "FunctionalBin" || colName == "WW" || colName == "SortLot" || colName == "SortWafer")
					{
						columns[colCount] = new TextLoader.Column(colName, DataKind.Single, colCount);
					}
					else
					{
						columns[colCount] = new TextLoader.Column(colName, DataKind.String, colCount);
					}

					colCount = colCount + 1;
                }


                columns[colCount] = new TextLoader.Column(labelColumnName, DataKind.String, colCount);

				var options = new TextLoader.Options
				{
					Separators = new[] { ',' },
					HasHeader = true,
					Columns = columns,
					AllowQuoting = true,
					AllowSparse = true
				};

				var loader = _mlContext.Data.CreateTextLoader(columns,
						hasHeader: true,
						separatorChar: ',');

                File.WriteAllText(testDataPath, content);

                var testData = loader.Load(testDataPath);

                ITransformer mlModel = _mlContext.Model.Load(modelPath, out var modelInputSchema);

                var output = mlModel.Transform(testData);


When calling the Transform, it is giving error "Could not find input column 'Features' (Parameter 'inputSchema')". My test data obviously has 27 columns just like the training data. 

How and why does the AutoML transformed the schema from 27 columns into the 13 columns? 

During predict, how do I convert the test data to match the schema created by the AutoML during training?
